velox/exec/tests/utils/TpchQueryBuilder.cpp

```diff
@@ -904,6 +906,123 @@ TpchPlan TpchQueryBuilder::getQ19Plan() const {
+
+  const std::string shipDateFilter = formatDateFilter(
+      "l_shipdate", lineitemSelectedRowType, "'1994-01-01'", "'1994-12-31'");
+
+  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+  core::PlanNodeId lineitemScanNodeId;
+  core::PlanNodeId partScanNodeId;
+  core::PlanNodeId supplierScanNodeId;
+  core::PlanNodeId partsuppScanNodeId;
+  core::PlanNodeId nationScanNodeId;
+
+  auto part = PlanBuilder(planNodeIdGenerator)
+                  .tableScan(
+                      kPart,
+                      partSelectedRowType,
+                      partFileColumns,
+                      {},
+                      "p_name like 'forest%'")
+                  .capturePlanNodeId(partScanNodeId)
+                  .planNode();
+
+  auto nation = PlanBuilder(planNodeIdGenerator)
+                    .tableScan(
+                        kNation,
+                        nationSelectedRowType,
+                        nationFileColumns,
+                        {"n_name = 'CANADA'"})
+                    .capturePlanNodeId(nationScanNodeId)
+                    .planNode();
+
+  auto partsuppJoinPart =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kPartsupp, partsuppSelectedRowType, partsuppFileColumns)
+          .capturePlanNodeId(partsuppScanNodeId)
+          .hashJoin(
+              {"ps_partkey"},
+              {"p_partkey"},
+              part,
+              "",
+              {"ps_partkey", "ps_suppkey", "ps_availqty"},
+              core::JoinType::kLeftSemi)
+          .planNode();
+
+  auto lineitemJoinPartsupp =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(
+              kLineitem,
+              lineitemSelectedRowType,
+              lineitemFileColumns,
+              {shipDateFilter})
+          .capturePlanNodeId(lineitemScanNodeId)
+          .partialAggregation(
+              {"l_partkey", "l_suppkey"}, {"sum(l_quantity) AS sum_qty"})
+          .localPartition({})
+          .finalAggregation()
+          .project({"l_partkey", "l_suppkey", "0.5 * sum_qty AS filter_qty"})
+          .hashJoin(
+              {"l_partkey", "l_suppkey"},
+              {"ps_partkey", "ps_suppkey"},
+              partsuppJoinPart,
+              "ps_availqty > filter_qty",
+              {"ps_suppkey"},
+              core::JoinType::kRight)
+          .planNode();
+
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kSupplier, supplierSelectedRowType, supplierFileColumns)
+          .capturePlanNodeId(supplierScanNodeId)
+          .hashJoin(
+              {"s_nationkey"},
+              {"n_nationkey"},
+              nation,
+              "",
+              {"s_name", "s_address", "s_suppkey"})
+          .hashJoin(
+              {"s_suppkey"},
+              {"ps_suppkey"},
+              lineitemJoinPartsupp,
+              "",
+              {"s_name", "s_address"},
+              core::JoinType::kLeftSemi)
```

Comment:
Like you mention in the sheet, right semi-join might be more useful here. Did you stick to this plan as it is what DuckDB has chosen ? If yes, then please add a comment here about it.

Reply:
Yes, this is consistent with the DuckDb plan, will update PR description, thanks!